features/macros.go

```diff
@@ -0,0 +1,206 @@
+	// EINVAL occurs when attempting to create a program with an unknown type.
+	// E2BIG occurs when ProgLoadAttr contains non-zero bytes past the end
+	// of the struct known by the running kernel, meaning the kernel is too old
+	// to support the given map type.
+	case errors.Is(err, unix.EINVAL), errors.Is(err, unix.E2BIG):
+		err = ebpf.ErrNotSupported
+
+	// EPERM is kept as-is and is not converted or wrapped.
+	case errors.Is(err, unix.EPERM):
+		break
+
+	// Wrap unexpected errors.
+	case err != nil:
+		err = fmt.Errorf("unexpected error during feature probe: %w", err)
+
+	default:
+		fd.Close()
+	}
+
+	macros.macroTypes[mt] = err
+
+	return err
+}
+
+func createMacroProbeAttr(mt MacroType) (*sys.ProgLoadAttr, error) {
+	var (
+		insns asm.Instructions
+		label string
+	)
+
+	switch mt {
+	case LargeInsn:
+		for i := 0; i < maxInsns; i++ {
+			insns = append(insns, asm.Mov.Imm(asm.R0, 1))
+		}
+		insns = append(insns, asm.Return())
+	case BoundedLoops:
+		label = "boundedLoop"
+		insns = asm.Instructions{
+			asm.Mov.Imm(asm.R0, 10),
+			asm.Sub.Imm(asm.R0, 1).Sym(label),
+			asm.JNE.Imm(asm.R0, 0, label),
+			asm.Return(),
+		}
+		// Manually adjust jump offset.
+		if err := adjustJumpOffset(insns, -2); err != nil {
+			return nil, err
+		}
+	case V2ISA:
+		label = "v2isa"
+		insns = asm.Instructions{
+			asm.Mov.Imm(asm.R0, 0).Sym(label),
+			asm.JLT.Imm(asm.R0, 0, label),
+			asm.Mov.Imm(asm.R0, 1),
+			asm.Return(),
+		}
+		// Manually adjust jump offset.
+		if err := adjustJumpOffset(insns, 1); err != nil {
+			return nil, err
+		}
+	case V3ISA:
+		label = "v3isa"
+		insns = asm.Instructions{
+			asm.Mov.Imm(asm.R0, 0).Sym(label),
+			asm.JLT.Imm32(asm.R0, 0, label),
+			asm.Mov.Imm(asm.R0, 1),
+			asm.Return(),
+		}
+		// Manually adjust jump offset.
+		if err := adjustJumpOffset(insns, 1); err != nil {
+			return nil, err
+		}
+	default:
+		return nil, fmt.Errorf("feature %d not yet implemented", mt)
+	}
+
+	buf := bytes.NewBuffer(make([]byte, 0, insns.Size()))
+	if err := insns.Marshal(buf, internal.NativeEndian); err != nil {
+		return nil, err
+	}
+
+	bytecode := buf.Bytes()
+	instructions := sys.NewSlicePointer(bytecode)
+
+	// Kernels before 5.0 (6c4fc209fcf9 "bpf: remove useless version check for prog load")
+	// require the version field to be set to the value of the KERNEL_VERSION
+	// macro for kprobe-type programs.
+	v, err := internal.KernelVersion()
+	if err != nil {
+		return nil, fmt.Errorf("detecting kernel version: %w", err)
+	}
+
+	return &sys.ProgLoadAttr{
+		ProgType:    sys.BPF_PROG_TYPE_SOCKET_FILTER,
+		Insns:       instructions,
+		InsnCnt:     uint32(len(bytecode) / asm.InstructionSize),
+		License:     sys.NewStringPointer("GPL"),
+		KernVersion: v.Kernel(),
+	}, nil
+}
+
+// adjustJumpOffset updates the first jump instruction in insns with the
+// given newJumpOffset.
+// If no jump instruction is found and error is returned.
+func adjustJumpOffset(insns asm.Instructions, newJumpOffset int16) error {
```

Comment:
Unfortunately, `insns.Marshal()` is not enough. 

In the usual workflow via `newProgramWithOptions()` this adjustment happens via `fixupJumpsAndCalls()`.

Reply:
@rgo3 @ti-mo why can't we use NewProgram from the features package?

Reply:
IIRC we decided on not using it because originally one of the goals was to use the `features` package in the main `ebpf` package to replace some of the current feature tests we have in there, I just never got to that during GSoC. So to not create any import cycles and as I didn't have to copy the entire function into the `features` package we ended up with what we have now. Additionally, at the time we decided that for my purposes it was enough to just use the `ProgLoad` syscall

Reply:
Thanks for the context. I think I'd opt for avoiding copypasta now, and deal with possible import cycles later.

Reply:
Could we avoid copypasta by exposing `fixupJumpsAndCalls()` somehow, maybe via `internal` and using it here? `NewProgram()` does not return all the errors from the syscall which we use for the features error interface if I see that correctly, so refactoring everything to use it  would be much more effort than having a shared function in internal that takes care of `fixupJumpsAndCalls()`.

Reply:
`fixupJumpsAndCalls` performs two actions: adjusting mumps and replacing bpf helper functions (for mordern kernels).
Splitting this function up and moving the adjusting of jumps into `internal` sounds like a good idea :+1: so it can be reused here in package `features` as well. wdyt?

Reply:
I can think of scenarios, where people hand craft their own instructions and with `Instructions.Marshal()` they get something different than they would expect. Therefore doing the jump fixup just in `Instructions.Marshal()` isn't a good idea, I think.
But having it in `asm` instead of `internal` sounds good to me. What about `Instructions.RewriteJumps(symbolOffsets map[string]asm.RawInstructionOffset)`?
If the given `symbolOffsets` is nil the offsets are calculated as it is happening right now. Otherwise the given offsets are used.

Reply:
Synced with @lmb in slack about this. Therefore updated the PR to use `Instructions.RewriteJumps()`.